src/client.rs

```diff
@@ -184,49 +187,75 @@ fn apply_init_message(
         stats.bytes += end_pos;
     }
 
-    let message_tick = bincode::deserialize_from(&mut cursor)?;
-    trace!("applying init message for {message_tick:?}");
-    world.resource_mut::<ServerInitTick>().0 = message_tick;
-    debug_assert!(cursor.position() < end_pos, "init message can't be empty");
-
-    apply_entity_mappings(world, params, &mut cursor)?;
-    if cursor.position() == end_pos {
-        return Ok(());
-    }
-
-    apply_despawns(world, params, &mut cursor, message_tick)?;
-    if cursor.position() == end_pos {
-        return Ok(());
-    }
-
-    apply_init_components(
-        world,
-        params,
-        ComponentsKind::Removal,
-        &mut cursor,
-        message_tick,
-    )?;
-    if cursor.position() == end_pos {
-        return Ok(());
+    let flags = ChangeMessageFlags::from_bits_retain(cursor.read_fixedint()?);
+    debug_assert!(!flags.is_empty(), "message can't be empty");
+
+    let message_tick = DefaultOptions::new().deserialize_from(&mut cursor)?;
+    trace!("applying change message for {message_tick:?}");
+    world.resource_mut::<ServerChangeTick>().0 = message_tick;
+
+    let last_flag = flags.last();
+    for (_, flag) in flags.iter_names() {
+        match flag {
+            ChangeMessageFlags::MAPPINGS => {
+                let len = apply_sized_array(&mut cursor, |cursor| {
+                    apply_entity_mapping(world, params, cursor)
+                })?;
+                if let Some(stats) = &mut params.stats {
+                    stats.mappings += len as u32;
+                }
+            }
+            ChangeMessageFlags::DESPAWNS => {
+                let len = if flag != last_flag {
+                    apply_sized_array(&mut cursor, |cursor| {
+                        apply_despawn(world, params, cursor, message_tick)
+                    })?
+                } else {
+                    apply_dyn_array(&mut cursor, |cursor| {
+                        apply_despawn(world, params, cursor, message_tick)
+                    })?
+                };
```

Comment:
I initially implemented it this way, but then I started using it in other places and quickly realized that calling `apply_array(false, ...)` is error-prone. I even fall for it myself :sweat_smile: I also considered using enum, but this will require the same `if`, this is why I went with two functions.

I can create a closure just for this function to reuse it inside. What do you think?

Reply:
Sure, closure is fine.